File: solutions/1034/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Mon Dec  6 22:00:35 2021
> Description:   
 ************************************************************************/
#include <queue> 
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

typedef pair<int, int> rc;

class Solution {
public:
    bool shouldAdd(const vector<vector<int>>& grid, int r, int c, int row, int col, vector<vector<int>>& visited) {
        // check base case
        if (r < 0 || c < 0 || r > grid.size()-1 || c > grid[0].size()-1) return false;
        else if (grid[r][c] != grid[row][col]) return false;
        else if (visited[r][c] != 0) return false;
        // need to prevent adding the same pair in queue
        visited[r][c] = 1;
        return true;
    }
    vector<vector<int>> colorBorder(vector<vector<int>>& grid, int row, int col, int color) {
        vector<vector<int>> visited(grid.size(), vector<int>(grid[0].size(), 0));
        queue<pair<int, int>> myQue;
        myQue.emplace(row, col);
        vector<int> dirR{-1, 1, 0, 0};
        vector<int> dirC{0, 0, -1, 1};
        visited[row][col] = 1;

        while (!myQue.empty()) {
            rc& p = myQue.front();
            myQue.pop();
            int r = p.first, c = p.second;

            // do bfs
            for (int k = 0; k < dirR.size(); ++k) {
                int x = r+dirR[k], y = c+dirC[k];
                if (shouldAdd(grid, x, y, row, col, visited))
                    myQue.emplace(x, y);
            }

            if (r == 0 || c == 0 || r == grid.size()-1 || c == grid[0].size()-1) {
                visited[r][c] = 2;
            }
            else {
                for (int k = 0; k < dirR.size(); ++k) {
                    int x = r+dirR[k], y = c+dirC[k];
                    if (grid[x][y] != grid[row][col]) {
                        visited[r][c] = 2;
                        break;
                    }
                }
            }
        }
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (visited[i][j] == 2) grid[i][j] = color;
            }
        }
        return grid;
    }
};

TEST(Test1034, SimpleTest) {
    Solution s; 
    vector<vector<int>> grid{{1,2,2,1,1,1,2},{2,1,1,1,2,2,1},{2,1,1,1,2,1,2},{2,2,1,1,1,1,1},{1,2,2,1,2,1,1}}; 
    int row = 4, col = 5, color = 1; 
    s.colorBorder(grid, row, col, color); 
    for (auto& r : grid) {
        for (auto& c : r)
            cout << c << " "; 
        cout << endl; 
    }
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}

File: solutions/1034/main.py
from typing import List 

class Solution:
    def shouldVisit(self, grid, row, col, oldColor, visited):
        if row < 0 or col < 0 or row > len(grid)-1 or col > len(grid[0])-1:
            # check if inside grid
            return False
        elif visited[row][col] != 0:
            # check if visited
            return False
        elif grid[row][col] != oldColor:
            return False
        else:
            return True

    def dfs(self, grid, row, col, oldColor, visited):
        # base case
        if not self.shouldVisit(grid, row, col, oldColor, visited):
            return
        else:
            visited[row][col] = 1

        # do color
        # up, down, left, right
        dirRow = [-1, 1, 0, 0]
        dirCol = [0, 0, -1, 1]

        if row == 0 or col == 0 or row == len(grid)-1 or col == len(grid[0])-1:
            # check if on border
            visited[row][col] = 2
        else:
            # check if adjacent to different color
            for i in range(len(dirRow)):
                if grid[row+dirRow[i]][col+dirCol[i]] != oldColor:
                    visited[row][col] = 2
                    break

        # dfs
        for i in range(len(dirRow)):
            self.dfs(grid, row+dirRow[i], col+dirCol[i], oldColor, visited)

    def colorBorder(self, grid: List[List[int]], row: int, col: int, color: int) -> List[List[int]]:
        visited = [[0] * len(grid[0]) for _ in range(len(grid))]
        oldColor = grid[row][col]
        self.dfs(grid, row, col, oldColor, visited)
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if visited[row][col] == 2:
                    grid[row][col] = color
        return grid

if __name__ == "__main__":
    grid = [[1,1,1],[1,1,1],[1,1,1]]
    row, col = 1, 1
    color = 2
    ans = [[2,2,2],[2,1,2],[2,2,2]]
    s = Solution()
    grid = s.colorBorder(grid, row, col, color)
    for row in range(len(grid)): 
        assert grid[row] == ans[row]
